### ISSUE TITLE
Shared framework & web.config configuration question

### ISSUE THREAD

**Issue by GuardRex** [aspnet/IISIntegration#125]
@Tratcher I thought something might be wrong with `dotnet cli`, but @davidfowl suggested that perhaps I just have my `web.config` improperly configured for the shared framework (on 2012R2).
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>
    <handlers>
      <add
        name="aspNetCore"
        path="*"
        verb="*"
        modules="AspNetCoreModule"
        resourceType="Unspecified"/>
    </handlers>
    <aspNetCore
        processPath="dotnet"
        arguments="run .\testshared.dll"
        stdoutLogEnabled="true"
        stdoutLogFile=".\logs\test"
        startupTimeLimit="3600"/>
  </system.webServer>
</configuration>
```
The `processPath` and `arguments` I set manually, because I thought that `dotnet run .\myapp.dll` was the way these shared framework apps were supposed to be run .... at least until `dotnet myapp.dll` works.

Getting a 502.3 (browser), "Process '0' failed to start." (OS App log), and a created but empty `AstNetCoreModule` log.

I used the `.NET Core Installer` from the `dotnet/cli` repo to get the shared framework on the server.

**Comment by GuardRex** [comment]
Probably will run with `dotnet testshared.dll`: `processPath="dotnet"` and `arguments="testshared.dll"`. My bad :monkey: ... but your `dotnet-publish-iis` tool obscures what will actually end up in the file when publishing samples (i.e., you can't see the actual command in sample `web.config` files in the `cli-samples` repo), and I'm not using the tool and using my own sample for testing with other app bits. It actually isn't explicitly stated in a `web.config` file that I've seen in the issues or docs.

**Comment by moozzyk** [comment]
publish-iis can't be run in this repo - at least for now (https://github.com/dotnet/cli/issues/1215). I believe the right mix is `processPath="%PROGRAMFILES\dotnet\dotnet,exe%" arguments=".\myapp.dll"`

**Comment by GuardRex** [comment]
Yes, except the app in question here is my own homebrew test app ... not a sample from a repo. 

It would be sufficient even if I don't use the dotnet command all by itself and hardcode the path, because what I'm doing right now is just working on the Publish to IIS doc. I just need the darn thing to run, but it was good to see the behavior when the path is wrong. 

I think you'll agree that post-RTM devs will make mistakes when they aren't using your excellent tooling. I'd like to see the Troubleshooting section of the doc cover a greater latitude of failure scenarios with more tips on what the dev should check before opening an issue or asking us on JabbR for help.

I'm getting back to this later today. Ill reopen if things :boom: explode :boom:.

**Comment by moozzyk** [comment]
I pointed it out only because I know for a fact that bare `dotnet` won't work. I don't expect that everyone will have/use tooling so yes, we will need to have samples that show you how things should be set up. (And I just tried a couple things and it :boom: https://github.com/aspnet/IISIntegration/issues/127 :boom:)

**Comment by GuardRex** [comment]
I'm glad u said so. Every tip helps. I'm going to see if the pic will
attach to the issue by email, because I'm landscaping right now because I
refuse to work in MVC4 and sometimes working with Core and landscaping have
something in common. :smile:

**Comment by GuardRex** [comment]
@moozzyk Well ... can't respond via E-mail to GH with an image. Ok. The landscaping actually went better than the testing. Here's what I have thus far ...

`processPath="dotnet" arguments="testshared.dll"` yields ...
```
Could not load host policy library from [testshared.dll]
```
`processPath="dotnet" arguments=".\testshared.dll"` yields ...
```
Could not load host policy library from 
[C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.0-rc3-24007-00]
This may be because the targeted framework [Microsoft.NETCore.App 1.0.0-rc3-24007-00] 
was not found.
```
... and that's not happening (at least) because this is what's actually there ...

![capture](https://cloud.githubusercontent.com/assets/1622880/14399758/908552c0-fdb5-11e5-9c35-02ce39286e65.PNG)

`1.0.0-rc2-23931` ... and I did have the package it's asking for locally, so I copied `1.0.0-rc3-24007-00` to the folder on the server. That didn't work.

Next, I tried to peg to RC2 going from ...
```json
"Microsoft.NETCore.App": {
    "type": "platform",
    "version": "1.0.0-*"
}
```
to
```json
"Microsoft.NETCore.App": {
    "type": "platform",
    "version": "1.0.0-rc2-*"
}
```
... and it came back with ...
```
Could not load host policy library from 
[C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.0-rc3-24007-00]
This may be because the targeted framework [Microsoft.NETCore.App 1.0.0-rc2-3002306] 
was not found.
```
The only thing that went well was that `dotnet` is on the PATH, so I didn't need to use the longer-ish `processPath="%PROGRAMFILES\dotnet\dotnet.exe%"`. It seems ok with just `processPath="dotnet"`.

I'm going to reopen right now, unless you have another issue I can track for getting a fix on this.

Any new news on this today (while I was under-shrub)?

**Comment by GuardRex** [comment]
Thought if I tried to sync the `Microsoft.NETCore.App` package better it might work. Published with -2296, and the .NET Core Installer for -2296 gave me a shared framework with `Microsoft.NETCore.App 1.0.0-rc2-3002338`. Published output from `"Microsoft.NETCore.App": { "type": "platform", "version": "1.0.0-rc2-*"}` resolved to `Microsoft.NETCore.App 1.0.0-rc2-3002306`, which was already on the server. Deploy and end up with ...

```
Could not load host policy library from 
[C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.0-rc2-3002306]
This may be because the targeted framework [Microsoft.NETCore.App 1.0.0-rc2-3002306] 
was not found.
```
... but ... the package is there ...
![capture](https://cloud.githubusercontent.com/assets/1622880/14401335/c18ded8c-fdd3-11e5-9de2-43023793729a.PNG)

idk ... Is the "host policy library" some other file that isn't here?

Shared framework is not functional; and with issues marked WIP and discussing the meta-package, I'll try again late next week.

**Comment by moozzyk** [comment]
Can you share your project?

**Comment by GuardRex** [comment]
@moozzyk 

https://github.com/GuardRex/sharedapp

I was going to wait until Thursday/Friday to give them time to gel a few things; but if you think I have a shot at it working today, then I'll try again later this afternoon with it.

**Comment by moozzyk** [comment]
I changed the NuGet.config file like this: 
https://github.com/GuardRex/sharedapp/compare/master...moozzyk:pawelka/nuget
and am able to run the app on:

```
PS C:\temp\repro\sharedapp> C:\Users\pawelka\AppData\Local\Microsoft\dotnet\dotnet.exe --version
1.0.0-beta-002279
```

**Comment by GuardRex** [comment]
@moozzyk Ok ... thanks. I'll give it a go again this evening and close this when it runs.

**Comment by GuardRex** [comment]
@Tratcher @moozzyk Working ... but ?

When I first loaded it up, I left the ...
```c#
.UseUrls("http://localhost:5001")
```
I received a `Error -4092 EACCES permission denied`, because (apparently) the port was in conflict with the binding in IIS. I removed the string, leaving ...
```c#
.UseUrls()
```
... but the exception persisted. It wasn't until I removed the `.UseUrls()` extension completely that it sprang to life.

What is the guidance here? Is the guidance for hosting in IIS never to use `.UseUrls()` and just set the port binding in IIS and leave it at that?

Also, I have not needed to use `.UseContentRoot(Directory.GetCurrentDirectory())` in any of my IIS-hosted apps. Since the `web.config` is in the content root now (app base), we really don't need this line for any reason, correct? ... even for the debugging scenario we were discussing before?

**Comment by Tratcher** [comment]
Right, UseUrls is for self-host, don't use it in IIS where the public binding is managed for you. 

In theory you can put UseUrls before UseIISIntegration and the value will be overwritten if the IIS environment variables are detected. This would enable the app to run in self-host and IIS.

**Comment by moozzyk** [comment]
What @Tratcher said.

With regards to `.UseContentRoot` - isn't it the same discussion we already had about views when developing the application? For IIS it was being overridden with an environment variable. This was removed when we moved the web.config to the app folder (https://github.com/aspnet/IISIntegration/commit/7a923d1d6fd105a3a7bb1bd992499716958f23bd#diff-3ffcd9a9b2c5202aadac411be113a50bL71) because it is no longer needed.

**Comment by GuardRex** [comment]
@moozzyk Have you seen this one? ... just cropped up with a fresh shared framework publish of my test app ...
```
Error loading hostpolicy 1.0.1-rc2-002296; interface mismatch between hostpolicy [32771] and 
hostfxr [120] Specifically, the structure of corehost_init_t has changed, do not know how to 
interpret it
```
I used the same `dotnet cli` version for both the publish and the .NET Core Installer for the server.
```
.NET Command Line Tools (1.0.0-rc2-002363)

Product Information:
 Version:     1.0.0-rc2-002363
 Commit Sha:  065e8352fc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64
```

**Comment by GuardRex** [comment]
@moozzyk Discussing this now in https://github.com/dotnet/cli/issues/2452

**Comment by moozzyk** [comment]
@GuardRex delete the $env:LocalAppData\Microsoft\Dotnet dir

**Comment by GuardRex** [comment]
@moozzyk Ok, but Senthil said this was part of a broader design issue to address.

I'll (hopefully) be returning to this this evening. I'll try your suggestion. Thanks.

**Comment by moozzyk** [comment]
@GuardRex - possibly. But if you need to unblock yourself now this is what you could do. (At least I had to do this yesterday after I moved to a later cli - not that the path was like that because it was in our build which installs dotnet there)

**Comment by GuardRex** [comment]
He said to manually set the version of `Microsoft.NETCore.App` in my `testshared.runtimeconfig.json` file of the published output. When I did that, it sprang right to life. I just set the version to whatever it was asking for, and I guess since that version was in the `shared` folder, it :sparkles: just worked :sparkles:.

Either way, I'll give it a go again this evening. Right now, I'm fighting through updating that .NET Core on Nano Server blog post ... The Azure folks really love to change their Azure Portal screens! Ouch for blog maintainers who use a lot of portal screenshots! :cry:

**Comment by GuardRex** [comment]
@moozzyk Nah ... still not working.

The app runs locally if I make the workaround change that Sunthil suggested last night, but moving the app to the server breaks it.

The module does establish a log file (empty). The browser immediately gets a 502.3. The Application log shows `Process '0' failed to start. Port = 29151, Error Code = '-2147024894'.`, which might be a file not found exception.

I tried setting the package version in `testshared.runtimeconfig.json` file to each of the versions available in the `C:\Program Files\dotnet\shared\Microsoft.NETCore.App` folder without success.

The path you mentioned above ... I can't find that on the server or on the local machine.

Any other ideas that I might try. It runs here on my Win10 box ... just not on the server. I did run the shared framework installer on the server of the same version as the `dotnet cli` that I installed, restored, and compiled with locally.

**Comment by GuardRex** [comment]
Shared is **working** for me now on -2394. :tada: :smile: Thanks @moozzyk  ..... now I hope the server gods grant minimal regressions :pray:. Thanks again for your help.

**Comment by LoisOrtiz** [comment]
Skincare Panama To keep your skin in good shape, be sure to exfoliate a minimum of three times per week. Try using a facial scrub. Use a moisturizing exfoliant if you have sensitive skin. You can get rid of dirt in pores if you exfoliate. The more you add exfoliating into your regimen, the nicer your skin will look.

https://vimeo.com/220189520
